packages/astro/src/assets/types.ts

```diff
@@ -0,0 +1,111 @@
+/* eslint-disable @typescript-eslint/ban-types */
+import { VALID_INPUT_FORMATS, VALID_OUTPUT_FORMATS } from './consts.js';
+import { ImageService } from './services/service.js';
+
+export type ImageQualityPreset = 'low' | 'mid' | 'high' | 'max' | (string & {});
+export type ImageQuality = ImageQualityPreset | number;
+export type InputFormat = (typeof VALID_INPUT_FORMATS)[number] | (string & {});
+export type OutputFormat = (typeof VALID_OUTPUT_FORMATS)[number] | (string & {});
+
+declare global {
+	// eslint-disable-next-line no-var
+	var astroAsset: {
+		imageService?: ImageService;
+		addStaticImage?: ((options: ImageTransform) => string) | undefined;
+		staticImages?: Map<ImageTransform, string>;
+	};
+}
+
+/**
+ * Type returned by ESM imports of images and direct calls to imageMetadata
+ */
+export interface ImageMetadata {
+	src: string;
+	width: number;
+	height: number;
+	format: InputFormat;
+}
+
+/**
+ * All the transformations possible to an image
+ */
+export type ImageTransform = {
+	src: ImageMetadata | string;
+	width?: number;
+	height?: number;
+	quality?: ImageQuality;
+	format?: OutputFormat;
+	[key: string]: any;
+};
+
+type WithRequired<T, K extends keyof T> = T & { [P in K]-?: T[P] };
+type ImageSharedProps<T> = T & {
+	/**
+	 * Width of the image, the value of this property will be used to assign the `width` property on the final `img` element.
+	 *
+	 * For local images, this value will additionally be used to resize the image to the desired width, taking into account the original aspect ratio of the image.
+	 *
+	 * **Example**:
```

Comment:
I think if you use the `@example` JSDoc tag, then the IDE will render the code block correctly

Reply:
Oh neat! I didn't know that plain markdown could work!